src/scout_apm/core/socket.py

```diff
@@ -23,33 +23,50 @@
 
 class CoreAgentSocket(threading.Thread):
     _instance = None
-    _run_lock = threading.Semaphore()
+    _run_lock = threading.Lock()
+    _instance_lock = threading.Lock()
 
     @classmethod
     def instance(cls, *args, **kwargs):
-        if cls._instance is None:
-            cls._instance = cls(args, kwargs)
-        elif cls._instance.running() is False:
-            del(cls._instance)
-            cls._instance = cls(args, kwargs)
-        return cls._instance
+        with cls._instance_lock:
+            if cls._instance is None:
+                cls._instance = cls(args, kwargs)
+                return cls._instance
+            elif cls._instance.running() is False:
+                cls._instance = cls(args, kwargs)
+                return cls._instance
+            else:
+                return cls._instance
```

Comment:
remove the last else, and remove the first two returns, then just have one return at the end outside of the if/elif